### PR TITLE
New keyboard with custom layout: handwired/twadlee/tp69

### PR DESCRIPTION
**Pull request by twadleigh**
## Description

This adds a new keyboard with a custom layout. I have verified that the firmware builds, loads, and works successfully on the one extant example of this keyboard.

Its default keymap is minimally functional, but I would like to get this merged before doing more work on keymaps so that I might be able to exploit the QMK configurator for that purpose.

~This PR also includes a trivial change to `ps2_mouse.c` to eliminate its unnecessary dependence on `avr` headers (and therefore make it usable with this keyboard).~ I'll save anything related to PS/2 to when I have a complete and working implementation based on the ChibiOS serial interface.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/handwired/twadlee/tp69/tp69.c [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+#include "tp69.h"
+#include "ps2.h"
+
+uint8_t ps2_error = PS2_ERR_NONE;
+
+void    ps2_host_init(void) {}
+uint8_t ps2_host_send(uint8_t data) {return 0;}
+uint8_t ps2_host_recv_response(void) {return 0;}
+uint8_t ps2_host_recv(void) {return 0;}
+void    ps2_host_set_led(uint8_t usb_led) {}
```
Comment (zvecr) [inline keyboards/handwired/twadlee/tp69/tp69.c:27]: Given this is entirely stubs, does `PS2_MOUSE_ENABLE` add any value at all?

Comment (twadleigh) [reply keyboards/handwired/twadlee/tp69/tp69.c:27]: Not yet :-) After I opened this PR I spent the rest of the weekend determining the right implementation appropriate for my Teensy LC. I think I made substantial headway on that. I intended for this PR to be focused on the keyboard only. I probably should have branched again before adding these stubs.

I'll back that last commit out of this branch.

Comment (twadleigh) [reply keyboards/handwired/twadlee/tp69/tp69.c]: @zvecr stubs removed on this branch.

Comment (zvecr) [reply keyboards/handwired/twadlee/tp69/tp69.c]: As a FYI, #6490 does a lot of the refactor required to get it to build on ChibiOS.

File: keyboards/handwired/twadlee/tp69/config.h [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+/* disable action features */
+//#define NO_ACTION_LAYER
+//#define NO_ACTION_TAPPING
+//#define NO_ACTION_ONESHOT
+
+/* disable these deprecated features by default */
+#ifndef LINK_TIME_OPTIMIZATION_ENABLE
+#    define NO_ACTION_MACRO
+#    define NO_ACTION_FUNCTION
+#endif
```
Comment (zvecr) [inline keyboards/handwired/twadlee/tp69/config.h:203]:
```suggestion
#define NO_ACTION_MACRO
#define NO_ACTION_FUNCTION
```

File: keyboards/handwired/twadlee/tp69/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+// Defines names for use in layer keycodes and the keymap
+enum layer_names { _BASE };
+
+// Defines the keycodes used by our macros in process_record_user
+enum custom_keycodes { QMKBEST = SAFE_RANGE, QMKURL };
+
```
Comment (zvecr) [inline keyboards/handwired/twadlee/tp69/keymaps/default/keymap.c:23]:
```suggestion

```

File: keyboards/handwired/twadlee/tp69/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+void matrix_scan_user(void) {
+
+}
+
+bool led_update_user(led_t led_state) {
+    return true;
+}
+*/
```
Comment (zvecr) [inline keyboards/handwired/twadlee/tp69/keymaps/default/keymap.c:60]:
```suggestion
```

File: keyboards/handwired/twadlee/tp69/tp69.c [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    return process_record_user(keycode, record);
+}
+
+bool led_update_kb(led_t led_state) {
+    // put your keyboard LED indicator (ex: Caps Lock LED) toggling code here
+
+    return led_update_user(led_state);
+}
+*/
```
Comment (zvecr) [inline keyboards/handwired/twadlee/tp69/tp69.c:51]:
```suggestion
```

File: keyboards/handwired/twadlee/tp69/rules.mk [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+# If you don't know the bootloader type, then you can specify the
+# Boot Section Size in *bytes* by uncommenting out the OPT_DEFS line
+# Otherwise, delete this section
+#   Teensy halfKay      512
+#   Teensy++ halfKay    1024
+#   Atmel DFU loader    4096
+#   LUFA bootloader     4096
+#   USBaspLoader        2048
+# OPT_DEFS += -DBOOTLOADER_SIZE=4096
+
```
Comment (zvecr) [inline keyboards/handwired/twadlee/tp69/rules.mk:26]:
```suggestion

```

File: keyboards/handwired/twadlee/tp69/tp69.h [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * The first section contains all of the arguments representing the physical
+ * layout of the board and position of the keys.
+ *
+ * The second converts the arguments into a two-dimensional array which
+ * represents the switch matrix.
+ */
+
+#define LAYOUT(k00, k01, k02, k03, k04, k05, k06, k07, k50, k51, k52, k53, k54, k55, k56, k57, /**/ k10, k11, k13, k14, k15, k16, k17, k60, k61, k62, k63, k64, k65, k66, k67, /*   */ k20, k21, k23, k24, k25, k27, /**/ k70, k71, k72, k73, k74, k75, k77, /*       */ k30, k32, k33, k34, k35, k36, k37, k81, k82, k83, k84, k85, k86, /*           */ k40, k41, k43, k44, k46, k47, k91, k92, k94, k95, k96, k97) \
+    { {k00, k01, k02, k03, k04, k05, k06, k07}, {k10, k11, KC_NO, k13, k14, k15, k16, k17}, {k20, k21, KC_NO, k23, k24, k25, KC_NO, k27}, {k30, KC_NO, k32, k33, k34, k35, k36, k37}, {k40, k41, KC_NO, k43, k44, KC_NO, k46, k47}, {k50, k51, k52, k53, k54, k55, k56, k57}, {k60, k61, k62, k63, k64, k65, k66, k67}, {k70, k71, k72, k73, k74, k75, KC_NO, k77}, {KC_NO, k81, k82, k83, k84, k85, k86, KC_NO}, {KC_NO, k91, k92, KC_NO, k94, k95, k96, k97}, }
```
Comment (zvecr) [inline keyboards/handwired/twadlee/tp69/tp69.h:31]:
```suggestion
#define LAYOUT( \
    k00, k01, k02, k03, k04, k05, k06, k07, k50, k51, k52, k53, k54, k55, k56, k57, \
    k10, k11, k13, k14, k15, k16, k17, k60, k61, k62, k63, k64, k65, k66, k67, \
    k20, k21, k23, k24, k25, k27, \
    k70, k71, k72, k73, k74, k75, k77, \
    k30, k32, k33, k34, k35, k36, k37, k81, k82, k83, k84, k85, k86, \
    k40, k41, k43, k44, k46, k47, k91, k92, k94, k95, k96, k97 \
) \
{ \
    {k00, k01, k02, k03, k04, k05, k06, k07}, \
    {k10, k11, KC_NO, k13, k14, k15, k16, k17}, \
    {k20, k21, KC_NO, k23, k24, k25, KC_NO, k27}, \
    {k30, KC_NO, k32, k33, k34, k35, k36, k37}, \
    {k40, k41, KC_NO, k43, k44, KC_NO, k46, k47}, \
    {k50, k51, k52, k53, k54, k55, k56, k57}, \
    {k60, k61, k62, k63, k64, k65, k66, k67}, \
    {k70, k71, k72, k73, k74, k75, KC_NO, k77}, \
    {KC_NO, k81, k82, k83, k84, k85, k86, KC_NO}, \
    {KC_NO, k91, k92, KC_NO, k94, k95, k96, k97} \
}
```

File: keyboards/handwired/twadlee/tp69/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+ */
+#include QMK_KEYBOARD_H
+
+// Defines names for use in layer keycodes and the keymap
+enum layer_names { _BASE };
+
+// Defines the keycodes used by our macros in process_record_user
+enum custom_keycodes { QMKBEST = SAFE_RANGE, QMKURL };
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {[_BASE] = LAYOUT(KC_GRV, KC_1, KC_2, KC_3, KC_4, KC_5, KC_6, KC_LGUI, KC_7, KC_8, KC_9, KC_0, KC_MINS, KC_EQL, KC_PGUP, KC_PGDN, /**/ KC_TAB, KC_Q, KC_W, KC_E, KC_R, KC_T, KC_LALT, KC_Y, KC_U, KC_I, KC_O, KC_P, KC_LBRC, KC_RBRC, KC_BSLS, /*   */ KC_ESC, KC_A, KC_S, KC_D, KC_F, KC_G, /**/ KC_H, KC_J, KC_K, KC_L, KC_SCLN, KC_QUOT, KC_ENT, /*       */ KC_LSFT, KC_Z, KC_X, KC_C, KC_V, KC_B, KC_LCTL, KC_N, KC_M, KC_COMM, KC_DOT, KC_SLSH, KC_RSFT, /*           */ KC_LCTL, KC_LGUI, KC_LALT, KC_BSPC, KC_MS_BTN1, KC_MS_BTN3, KC_MS_BTN2, KC_SPC, KC_RALT, KC_RGUI, KC_APP, KC_RCTL)};
```
Comment (zvecr) [inline keyboards/handwired/twadlee/tp69/keymaps/default/keymap.c:24]:
```suggestion
const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
    [_BASE] = LAYOUT(
        KC_GRV, KC_1, KC_2, KC_3, KC_4, KC_5, KC_6, KC_LGUI, KC_7, KC_8, KC_9, KC_0, KC_MINS, KC_EQL, KC_PGUP, KC_PGDN,
        KC_TAB, KC_Q, KC_W, KC_E, KC_R, KC_T, KC_LALT, KC_Y, KC_U, KC_I, KC_O, KC_P, KC_LBRC, KC_RBRC, KC_BSLS,
        KC_ESC, KC_A, KC_S, KC_D, KC_F, KC_G, KC_H, KC_J, KC_K, KC_L, KC_SCLN, KC_QUOT, KC_ENT,
        KC_LSFT, KC_Z, KC_X, KC_C, KC_V, KC_B, KC_LCTL, KC_N, KC_M, KC_COMM, KC_DOT, KC_SLSH, KC_RSFT,
        KC_LCTL, KC_LGUI, KC_LALT, KC_BSPC, KC_MS_BTN1, KC_MS_BTN3, KC_MS_BTN2, KC_SPC, KC_RALT, KC_RGUI, KC_APP, KC_RCTL
    )
};
```

File: keyboards/handwired/twadlee/tp69/keymaps/default/config.h [hunk 9]
```diff
@@ -0,0 +1,19 @@
+/* Copyright 2020 Tracy Wadleigh
```
Comment (noroadsleft) [inline keyboards/handwired/twadlee/tp69/keymaps/default/config.h:1]: This file can safely be deleted.

File: keyboards/handwired/twadlee/tp69/readme.md [hunk 10]
```diff
@@ -0,0 +1,17 @@
+# Twadlee TP69
+
+![tp69](https://i.imgur.com/sC1qmJS.png)
+69-key keyboard having a Trackpoint unit under an additional 70th keycap.
```
Comment (noroadsleft) [inline keyboards/handwired/twadlee/tp69/readme.md:4]: Adding another line break here improves the readability/visibility of the description.

```suggestion

A 69-key keyboard having a Trackpoint unit under an additional 70th keycap.
```

File: keyboards/handwired/twadlee/tp69/tp69.h [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * The second converts the arguments into a two-dimensional array which
+ * represents the switch matrix.
+ */
+
+#define LAYOUT( \
+    k00, k01, k02, k03, k04, k05, k06, k07, k50, k51, k52, k53, k54, k55, k56, k57, \
+    k10, k11, k13, k14, k15, k16, k17, k60, k61, k62, k63, k64, k65, k66, k67, \
+    k20, k21, k23, k24, k25, k27, \
+    k70, k71, k72, k73, k74, k75, k77, \
```
Comment (noroadsleft) [inline keyboards/handwired/twadlee/tp69/tp69.h:34]:
```suggestion
    k20, k21, k23, k24, k25, k27, k70, k71, k72, k73, k74, k75, k77, \
```

File: keyboards/handwired/twadlee/tp69/keymaps/default/config.h (hunk 9 above)
Comment (twadleigh) [reply keyboards/handwired/twadlee/tp69/keymaps/default/config.h]: Done.

**Comment by twadleigh** [comment]
I have accepted all the suggestions, squashed, and rebased on `master`.

File: keyboards/handwired/twadlee/tp69/config.h [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+ * ROWS: AVR pins used for rows, top to bottom
+ * DIODE_DIRECTION: COL2ROW = COL = Anode (+), ROW = Cathode (-, marked on diode)
+ *                  ROW2COL = ROW = Anode (+), COL = Cathode (-, marked on diode)
+ */
+/*                        19  17  16  15  14   2   8   6   7   9 */
+#define MATRIX_ROW_PINS \
+    { B2, B1, B0, C0, D1, D0, D3, D4, D2, C3 }
+/*                        20  21  18    0    1   3   4   5 */
+#define MATRIX_COL_PINS \
+    { D5, D6, A4, B16, B17, A1, A2, D7 }
```
Comment (fauxpark) [inline keyboards/handwired/twadlee/tp69/config.h:48]:
```suggestion
#define MATRIX_ROW_PINS { B2, B1, B0, C0, D1, D0, D3, D4, D2, C3 }
/*                        20  21  18    0    1   3   4   5 */
#define MATRIX_COL_PINS { D5, D6, A4, B16, B17, A1, A2, D7 }
```

File: keyboards/handwired/twadlee/tp69/rules.mk [hunk 13]
```diff
@@ -0,0 +1,26 @@
+# MCU name
+MCU = MKL26Z64
+USE_CHIBIOS_CONTRIB = yes
+
+OPT_DEFS += -DCORTEX_ENABLE_WFI_IDLE=TRUE
```
Comment (fauxpark) [inline keyboards/handwired/twadlee/tp69/rules.mk:5]:
```suggestion

```

File: keyboards/handwired/twadlee/tp69/rules.mk [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
+HD44780_ENABLE = no         # Enable support for HD44780 based LCDs
+PS2_MOUSE_ENABLE = no
+PS2_USE_USART = no
```
Comment (fauxpark) [inline keyboards/handwired/twadlee/tp69/rules.mk:26]:
```suggestion
PS2_USE_USART = no

USE_CHIBIOS_CONTRIB = yes
OPT_DEFS += -DCORTEX_ENABLE_WFI_IDLE=TRUE
```

File: keyboards/handwired/twadlee/tp69/rules.mk (hunk 13 above)
Comment (twadleigh) [reply keyboards/handwired/twadlee/tp69/rules.mk:5]: Here I was following `keyboards/handwired/onekey/teensy_lc`: https://github.com/qmk/qmk_firmware/blob/4c4009c0fd3e55c2eacde302ef08e56a3ff2ab60/keyboards/handwired/onekey/teensy_lc/rules.mk#L6

Teensy LC support requires ChibiOS-Contrib.

Comment (fauxpark) [reply keyboards/handwired/twadlee/tp69/rules.mk]: These suggestions just move it to the bottom.

Comment (twadleigh) [reply keyboards/handwired/twadlee/tp69/rules.mk]: Didn't see the second half of the diff.

**Comment by twadleigh** [comment]
I gave it one last (I hope?) squash/rebase.

**Comment by twadleigh** [comment]
> Looks good. Not sure about the inclusion of the KLE JSON though.

@fauxpark: I figured it would be good practice to include the file from which I generated `info.json`. However, I will remove it in a future PR, since you judge it unnecessary.

**Comment by twadleigh** [comment]
*bump*

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**